Change is_constantt logic to enumerate not-constant expressions

tautschnig opened the pull request:
Rather than adding more-and-more expressions to the enumeration just call out those that are _not_ constant ((nondet) symbols and side-effect expressions as well as any access that may be out of bounds).


<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Change is_constantt logic to enumerate not-constant expressions

Rather than adding more-and-more expressions to the enumeration just
call out those that are _not_ constant ((nondet) symbols and side-effect
expressions as well as any access that may be out of bounds).

With the additional const-ness we need to make Array_UF22 a bit more
complicated (less constant) to continue to demonstrate the problem.

kroening commented:
I am starting to get worried about terminology. The word 'constant' can be interpreted in a logic sense (nullary functions), and considering various program semantics. In this case, the object size is definitively something that can change.

My suggestion would be to rename this to "forward-propagate" and move it to symex, because this is what this code does.

tautschnig commented:
> I am starting to get worried about terminology. The word 'constant' can be interpreted in a logic sense (nullary functions), and considering various program semantics. In this case, the object size is definitively something that can change.
> 
> My suggestion would be to rename this to "forward-propagate" and move it to symex, because this is what this code does.

There are other places that use it, such as the C front-end. What I have now done, however, is change the logic in this code. It might actually end up being even more controversial.

kroening commented:
> There are other places that use it, such as the C front-end. What I have now done, however, is change the logic in this code. It might actually end up being even more controversial.

In C, constant could mean `const` (can't assign, but might change) or `constexpr` (known at compile time). I strongly suspect that the symex engine wants neither one of these.

tautschnig commented:
> In C, constant could mean `const` (can't assign, but might change) or `constexpr` (known at compile time). I strongly suspect that the symex engine wants neither one of these.

Here is an overview of how we derive from `is_constantt` (which is pretty close to where we use it):

```
$ git grep is_constantt | grep class
src/analyses/constant_propagator.cpp:class constant_propagator_is_constantt : public is_constantt
src/ansi-c/c_typecheck_expr.cpp:class is_compile_time_constantt : public is_constantt
src/goto-instrument/havoc_utils.h:class havoc_utils_is_constantt : public is_constantt
src/goto-symex/goto_symex_is_constant.h:class goto_symex_is_constantt : public is_constantt
src/util/expr_util.h:class is_constantt
```

Perhaps we need to review all these names? Should the base class be renamed from `is_constantt` to `can_forward_propagatet`? I think `is_compile_time_constantt` is pretty accurate. The other three likely should get `forward_propagate` in their name.

kroening commented:
Yes, we need to review the naming, and update some of the comments.

Also, it bugs me that say the C front-end will declare something like a byte-update to be a compile-time constant even though the concept doesn't even exist. It may make sense to use specialised copies in some cases, even if that implies some duplication.

kroening commented:
The name `can_forward_propagatet` fits for the base class with the sole exception of the use in the C front-end for "compile-time constant". I feel that if that usage gets split out, things are much cleaner.

tautschnig committed:
Rename is_constantt (and it's children) to can_forward_propagatet

There is too much ambiguity in what could be called a "constant," and we
really mean "this expression can be forward propagated."

tautschnig committed:
Make is_compile_time_constantt independent of is_constantt

Compile-time constants can only be ones that actually appear in the
front-end. In a future commit is_constantt will be further disambiguated
and be renamed to can_forward_propagatet.

tautschnig commented:
> The name `can_forward_propagatet` fits for the base class with the sole exception of the use in the C front-end for "compile-time constant". I feel that if that usage gets split out, things are much cleaner.

I have now done the renaming and separated out the `is_compile_time_constantt` (in individual commits).
